### PR TITLE
Automatic Document Indexing Batches

### LINKED ISSUE #13880: Add Document Auto Batching Functionality
Add document auto batching functionality that will allow for a customer to pump documents into the object and it will implicitly handle generating and sending document batches. In addition to sending batches it will automatically handle retrying retriable documents that failed their operation and will dump documents that have completely failed.

There may be additional functionality to offload remaining documents in the batching object when the application is shutdown.

### PR DESCRIPTION
**Pull request by alzimmermsft**
Fixes #13880 

This PR adds the ability to have automatic batching via the `SearchIndexBatchingClients`. The batching client is constructed using two parameters to determine handling for automatically sending batches. First is a batching scheduler which tracks the window of time between two documents being inserted into the batch, if this value isn't set the client won't use automatic batch sending. Second is a batch size trigger, once the number of documents in the batch reaches or exceed the limit the batch will be sent only if automatic batch sending is enabled. Additionally, the batch client will leverage a new interface of `DocumentPersister` which will persist documents in the batch queue and succeeded or failed documents, this functionality is optional and is meant to offer a consistent store outside of memory for the client. The batch client will maintain three active queues of the pending documents to be index, documents that have completed indexing, and documents that have failed indexing with a non-retryable status code. If a batch is sent that exceeds the size limit of the Search service the batch will be split into two smaller batches and sent again, this process will repeat until a batch size of one and then at that time the operation will fail as the single document is too large for the service to accept.

### CODE REVIEW

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchAsyncClient.java [hunk 1]
```diff
@@ -164,6 +166,41 @@ public String getEndpoint() {
...
+    public SearchIndexBatchingAsyncClient getIndexDocumentBatchingAsyncClient() {
+        return getIndexDocumentBatchingAsyncClient(null, null, null, null);
+    }
+
+    /**
+     * Creates a {@link SearchIndexBatchingAsyncClient} used to index documents for the Search index associated with
+     * this {@link SearchAsyncClient}.
+     *
+     * @param autoFlush Flag determining whether the batching client will automatically flush its document batch. If
+     * {@code null} is passed this will be set to {@code true}.
```
Comment (mssfang) [inline sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchAsyncClient.java:188]: https://blog.joda.org/2012/11/javadoc-coding-standards.html#:~:text=Do%20not%20use%20%40code%20for,and%20adds%20no%20real%20value.

Comment (alzimmermsft) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchAsyncClient.java]: Updated

**Review by mssfang — APPROVED** [review]

**Pull request merged**